Initial MysqlDumpBundle commit

cH40z-Lord opened the pull request.

apfelbox commented:
Please add docs (regarding installation and usage) in `README.md`

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+    {
+        $this
+            ->setName('becklyn:db:dump')
+            ->setDescription('Dumps the configured or given Databases using \'mysqldump\' to .sql files.')
```
If you use double quotes, you can omit the escaping. Also: "databases" (with "d")

``` php
->setDescription("Dumps the configured or given databases using 'mysqldump' to .sql files.")
```

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+                'databases',
+                'd',
+                InputOption::VALUE_REQUIRED,
+                'If provided it overrides the values from the <info>config.yml</info>. Multiple Database names are separated by <info>comma (,)</info>',
```
The overriding should be mentioned in the docs, not here IMO.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+                null
+            )
+            ->addOption(
+                'yes',
```
`--force` would be a better name, I think.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+        $backupPath        = $input->getOption('path');
+        $suppressPrompt    = $input->getOption('yes');
+
+        $databases = null;
```
Not needed.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+            // 1) Use 'becklyn_mysqldump:databases' when set
+            // 2) else fallback to the databases from 'doctrine:dbal:connections'
+            $databases = $this->getDatabases();
+        }
```
This whole section seems unclear.
I would have implemented it like this (pseudo code):

``` php
$databases = $this->getDatabaseConnectionInfo($this->getDatabaseNamesToDump($selectedDatabases));


// ...

public function getDatabaseNamesToDump (array $cliArgument)
{
    // prefer database names as CLI argument ...
    if (!empty($cliArgument))
    {
        return $cliArgument;
    }

    // ... over configuration in config.yml ...
    if (!empty($databases = $this->getContainer()->get("becklyn_mysql_dump.configuration")->getConfig("databases")))
    {
        return $databases;
    }

    // ... over just dumping all associated databases.
    return $this->getAllDatabaseNames();
}
```

This makes the hierarchy clearer.

apfelbox replied:
Also we should probably deal with databases that either don't exist or are not configured as symfony-related.

If we want to allow dumping them, we need a way to explicitly add database user and password to them (probably defaulting to the one from the default connection).

apfelbox replied:
One note about the code example: you need to bump the required PHP version to 5.5, if you use the expression in `empty()`

apfelbox commented:
We should pull the dumping logic out into a custom service. That way we aren't completely bound to the CLI.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+
+        $output->writeln('<info>»</info> Found connection data for the following databases:');
+        $output->writeln('Database(s): <info>' . implode(',', array_keys($databases)) . '</info>');
+        $output->writeln("Backup directory: <info>$backupPath</info>");
```
Would it be better to draw a table?

| Database | Backupfile |
| --- | --- |
| example | app/var/example.sql |
| example2 | app/var/example2.sql |
| example3 | app/var/example3.sql |

etc...

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,250 @@
+        {
+            if (in_array($connection->getDatabase(), $databaseNames))
+            {
+                $connectionData[$connection->getDatabase()] = [
```
Isn't there a special class for it? You should create one and use it, instead of just simple arrays.

apfelbox commented on DependencyInjection/BecklynMysqlDumpExtension.php:
```diff
@@ -0,0 +1,32 @@
+
+        // Add the actual MySQL Dump configuration to the ConfigurationService
+        $definition = $container->getDefinition('becklyn_mysql_dump.configuration');
+        $definition->replaceArgument(0, $config);
```
I would split that into two arguments.

apfelbox commented on Resources/config/services.yml:
```diff
@@ -0,0 +1,4 @@
+services:
+    becklyn_mysql_dump.configuration:
+        class: Becklyn\MysqlDumpBundle\Service\ConfigurationService
+        arguments: [null]
```
List form, please

apfelbox commented on Service/ConfigurationService.php:
```diff
@@ -0,0 +1,43 @@
+     *
+     * @param $configuration
+     */
+    public function __construct ($configuration)
```
You should make that smarter. Use two parameters and do sanitization of them right here.

apfelbox replied:
Also use custom getters with proper PHPDoc comments.

cH40z-Lord commented:
Updated PR to reflect changes due to feedback.

Most notable changes are:
- the usage of connection names instead of database names. this prevents the edge case when you have multiple connections to different databases with the same name. they now no longer override each other's backup.
- moved the actual dump logic inside a dedicated service
- fancier output including table overview
- raised minimum requirements: now PHP 5.5 and Symfony 2.5

**Note**: I'm going to add a README.md as soon as we're satisfied with the current design of the MysqlDumpBundle and all of its features.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+                'path',
+                'p',
+                InputOption::VALUE_REQUIRED,
+                'The folder path where the .sql file will be saved. Defaults to %kernel.root%/var/db_backups/ if not specified otherwise.',
```
`"The folder path where the .sql file will be saved. Defaults to '%kernel.root%/var/db_backups/'.",`

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+    protected function execute (InputInterface $input, OutputInterface $output)
+    {
+        $dumper       = $this->getContainer()->get('becklyn_mysql_dump.services.dump');
+        $dumperConfig = $this->getContainer()->get('becklyn_mysql_dump.services.configuration');
```
Use a better (or no) prefix instead of `services.`

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+        $formatter = $this->getHelper('formatter');
+
+        // Print headline
+        $headline = $formatter->formatBlock(array('', '  Becklyn MySQL Database Dumper', ''), 'comment');
```
Use the short array syntax `[ /* ... */ ]`.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+        $this->getHelper('table')
+             ->setHeaders(array('Database', 'Connection', 'Backup file'))
+             ->setRows($tableRows)
+             ->render($output);
```
Extract the complete table data into its own method.

You don't need to pass the formatter helper, as you can easily fetch it again.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+             ->render($output);
+
+        // Check if any connections could not be resolved and then print an error to abort.
+        if ($unresolvedConnections)
```
If you extract the table into its method, you won't have this info available.

Just use `in_array(null, $connections, true)` now instead.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+        if (!$input->getOption('force'))
+        {
+            $output->writeln('');
+            $question = new ConfirmationQuestion('<question>Would you like to backup now? [Yn]</question>', true);
```
I would use `Would you like to start the backup now?`, what do you think?

keichinger replied:
Without the [Yn]?

I find them somewhat useful so you know what you need to enter. Also the capital letter shows which option it defaults to if you hit Enter. From what I see around in the Unix world it's kind of a standard (or maybe just a convention?).

In the end I wouldn't mind dropping the [Yn] part.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+
+            if (!$this->getHelper('question')->ask($input, $output, $question))
+            {
+                $output->writeln('<info>Exiting</info>. No files were written.');
```
Use `Aborting.` instead of `Exiting.`

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+            foreach ($connections as $connectionName => $connectionData)
+            {
+                $dumper->dump($output, $connectionName, $connectionData, $dbBackupPaths[$connectionName]);
+            }
```
Do you want to abort all backups or keep them rolling, although one or more did have an exception? It's probably better to have rather more backups than less?

keichinger replied:
The only exception we're currently handling/catching is the one that is thrown when the directory couldn't be created. In this case no backup can be created cause they don't have individual backup paths (which might be a good idea for the future btw) and all use the same, which couldn't be created.

For the future if we need to handle more edge cases we might move the try-catch inside the loop, if appropriate.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+                $dumper->dump($output, $connectionName, $connectionData, $dbBackupPaths[$connectionName]);
+            }
+        }
+        catch (AccessDeniedException $e)
```
Please use custom exceptions.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+                    ''
+                ],
+                'error'
+            );
```
Use can use the short form, if you want to.

``` php
$output->writeln(
    $formatter->formatBlock([
        '',
        '  An error occurred during backup creation:  ',
        "  {$e->getMessage()} ",
        ''
    ], 'error')
);
```

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,190 @@
+            $output->writeln($error);
+        }
+
+        $output->writeln("\n<info>»» Backup completed.</info>");
```
You previously used empty `writeln("")` calls.
You should use them here, too:

``` php
$output->writeln("");
$output->writeln("<info>»» Backup completed.</info>");
```

apfelbox commented on DependencyInjection/BecklynMysqlDumpExtension.php:
```diff
@@ -0,0 +1,32 @@
+ * This is the class that loads and manages your bundle configuration
+ *
+ * To learn more see {@link http://symfony.com/doc/current/cookbook/bundles/extension.html}
+ */
```
``` php
/**
 * @inheritdoc
 */
```

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,156 @@
+class DatabaseConfigurationService extends ContainerAware
+{
+    /**
+     * The MysqlDumpBundle Configuration (config.yml)
```
Obvious, that it comes from `config.yml` (as it's configuration)

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,156 @@
+     * @param Registry $doctrine
+     * @param array    $configuration
+     */
+    public function __construct (Registry $doctrine, array $configuration)
```
Pass `$configuration` as separate parameters and validate them.

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,156 @@
+        {
+            // Create an associative array with connection name => null as fallback.
+            // With each connection we can resolve we replace the value with the actual connection object
+            $connectionData = array_combine($connectionNames, array_fill(0, count($connectionNames), null));
```
Nice. :smiley:

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,156 @@
+            // or if the connectionNames array is empty, which means every connection will be included
+            if (in_array($name, $connectionNames) || empty($connectionNames))
+            {
+                $connectionData[$name] = $connection;
```
Hmm, I would probably use a custom object and add the resolved backup path to it, too.

apfelbox commented on Service/MysqlDumpService.php:
```diff
@@ -0,0 +1,78 @@
+
+class MysqlDumpService
+{
+
```
Remove empty line.

apfelbox commented on Service/MysqlDumpService.php:
```diff
@@ -0,0 +1,78 @@
+
+
+use Doctrine\DBAL\Connection;
+use Symfony\Component\Console\Output\OutputInterface;
```
Your Dump Service should have no connection to the console.

apfelbox commented on Service/MysqlDumpService.php:
```diff
@@ -0,0 +1,78 @@
+        else
+        {
+            $fileSystem->remove($backupPath);
+            $output->writeln('<error>failed</error>');
```
Throw an exception.

apfelbox commented on Service/MysqlDumpService.php:
```diff
@@ -0,0 +1,78 @@
+
+        return $success;
+    }
+}
```
We should think about how to better use it and extract the console component from it.

apfelbox commented on composer.json:
```diff
@@ -0,0 +1,21 @@
+  ],
+  "require":      {
+    "php": ">=5.5.0",
+    "symfony/framework-bundle": "~2.5"
```
Do you really need the framework bundle here or just the separate components? You should check that and just include them.

Seems like a waste to require the complete framework here.

apfelbox commented on composer.json:
```diff
@@ -0,0 +1,21 @@
+    }
+  ],
+  "require":      {
+    "php": ">=5.5.0",
```
`"~5.5"`

apfelbox commented on composer.json:
```diff
@@ -0,0 +1,21 @@
+  "homepage":     "https://github.com/Becklyn/BecklynMysqlDumpBundle",
+  "authors":      [
+    {
+      "name":     "Becklyn GmbH",
```
We only want to use `"Becklyn"` from now on.

apfelbox commented on .gitignore:
```diff
@@ -0,0 +1,2 @@
+composer.lock
```
This file should never be ignored.

apfelbox commented on Command/DumpCommand.php:
```diff
@@ -0,0 +1,224 @@
+     * @param OutputInterface $output
+     * @param Exception       $e
+     */
+    protected function printDumpException (OutputInterface $output, Exception $e)
```
You should use the type `MysqlDumpException` directly, instead of `Exception`.

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,150 @@
+     * ConfigurationService constructor.
+     *
+     * @param Registry $doctrine
+     * @param array    $connectionIdentifiers
```
You could use `string[]` here.

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,150 @@
+     * Returns all database connections that should be backed up by searching the configuration:
+     *  1) Uses, if provided, the arguments passed via CLI
+     *  2) Searches the config.yml for the key 'becklyn_mysql_dump:databases'
+     *  3) Searches the config.yml for the connections set up under 'doctrine:dbal:connections'
```
These config options are separated via `.`, like: `doctrine.dbal.connections`

apfelbox replied:
```
3) Searches the config.yml for the connections set up under 'doctrine:dbal:connections'
```

-> 

```
3) Uses all database connections as configured for doctrine.
```

apfelbox commented on Service/DatabaseConfigurationService.php:
```diff
@@ -0,0 +1,150 @@
+    /**
+     * Returns a list of all databases that are configured under 'becklyn_mysql_dump:databases'
+     *
+     * @return array An associative array of identifier => DatabaseConnection
```
You don't need the explicit `identifier` here, as this is already included in `DatabaseConnection`.

Also use explicit array types instead of `array`, if the arrays are implicitly typed. In this case: `DatabaseConnection[]` -- you will get proper type suggestions in your IDE.

keichinger replied:
I once had DatabaseConnection[] instead of the associative array which we're using now until the moment I realised I need some way to list DatabaseConnections I couldn't resolve. If I go for DatabaseConnection[] only I would need to create dummy objects and test whether they could be resolved or not.

The only way we could achieve this is by introducing yet another field to the DatabaseConnection class to flag whether the connection could be resolved. I'm just not sure how clean it is to create a new DatabaseConnection object which doesn't have any values except its identifier and the flag to mark it unresolved/incomplete.

I'm open for suggesstions.

apfelbox replied:
You are right, we leave the map instead of a plain array.

But please update the typehint to identify the more concrete types:

The proper type would either be `(DatabaseConnection|null)[]` or `array<DatabaseConnection|null>`. You can opt for  `DatabaseConnection[]`, too, as the correct ones doesn't seem to be supported by PhpStorm yet. :(

apfelbox commented on Service/MysqlDumpService.php:
```diff
@@ -0,0 +1,176 @@
+        // We need to determine ourselves whether mysqldump has raised an error
+        // as its return code is unreliable due to the fact that it's always returning 0
+        $errorOut = trim($process->getErrorOutput());
+        $success  = (preg_match('/Got error:.*/', $errorOut) !== 1);
```
`stripos(errorOut, "Got error:") !== false` is way faster here.

apfelbox commented on Service/MysqlDumpService.php:
```diff
@@ -0,0 +1,176 @@
+            // Ensure that MySQLi is throwing proper exceptions
+            $driver->report_mode = MYSQLI_REPORT_ALL;
+
+            $mysqli = new mysqli($connection->getHost(), $connection->getUsername(), $connection->getPassword(), $connection->getDatabase(), $connection->getPort());
```
I would just use PDO to check for the connection. That way we ensure forward compatibility if we want to support other databases in the future.

keichinger replied:
My impression was that we would simply create dedicated *DumpService classes for each database type we add in the future anyway, which could also handle the connection validation differently. 

I don't mind which way we implement it as long as we decide on one :)

apfelbox replied:
We can leave it as it is for now, but normally PDO is the preferred database abstraction engine IIRC.

apfelbox commented:
One last thing: as we are implementing possibilities to dumping other databases as well, please rename the bundle, package etc.. just to `BecklynDatabaseDumpBundle`.

apfelbox commented:
You can leave the `:db:` abbreviation in the command, though.

apfelbox commented:
(I will update the repo name as soon as everything is merged)

apfelbox commented:
:balloon:
